new devicetree.h API

mbolivar opened the pull request:
Add a new devicetree.h API. This makes the current macros deprecated but keeps them around. The new API uses macros which are **not** generated, hiding the details of the generated macros from the user.

This is inspired by https://github.com/zephyrproject-rtos/zephyr/issues/22964#issuecomment-591785043 but "reverses" the capitalization: the names from devicetree remain lowercase. See the documentation changes for what I mean.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/22964
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/22555
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/22554

Per @galak and @pabigot:

Fixes: #21369
Fixes: #19285

TODO list:

- [x] fix documentation
- [x] pass CI (some linter checks must be ignored)
- [x] tests for all DT macros
- [x] list all the property suffixes in one place
- [x] cs-gpios helpers
- [x] finish fixing up howtos.rst
- [x] settle on an approach to phandles
- [x] reg and irq name macros
- [x] DT_NODE_HAS_COMPAT
- [x] DT_NUM_REGS, DT_NUM_IRQS
- [x] generate a define for number instances ie DT_N_INST_<compat>_NUM
- [x] IRQ cell existence checks

Out of scope for this PR, to be addressed later:

- flash partitions

mbolivar-nordic committed:
devicetree.h: add accessor API for nodes and properties

This is joint work with Kumar Gala (see signed-off-by).

Add helper macros which abstract the "true names" of each of the four
types of node identifier we intend to support (e.g. DT_ALIAS(),
DT_INST()).

These can be passed to a new DT_PROP() macro which can be used to read
the value of a devicetree property given a node identifier from one of
these four other macros, and the as-a-c-token name of the property.
Add other accessor macros and tests as well.

Add some convenience APIs for writing device drivers based on instance
numbers as well. Drivers can "#define DT_DRV_COMPAT driver_compatible"
at the top of the file, then utilize these DT_INST_* macros to access
various property defines.

For example, the uart_sifive driver can do:

  #define DT_DRV_COMPAT sifive_uart0

Then use DT_INST macros like:

  .port         = DT_INST_REG_ADDR(0),
  .sys_clk_freq = DT_INST_PROP(0, clock_frequency),

For convenience working with specific hardware, also add:

  <devicetree/gpio.h>
  <devicetree/adc.h>
  <devicetree/spi.h>

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic committed:
scripts: dts: write new gen_defines.py

This is joint work with Kumar Gala (see signed-off-by).

This supports a new devicetree macro syntax coming. It's not really
worth mixing up the old and the new generation scripts into one file,
because:

- we aim to remove support for the old macros at some point, so it
  will be cleaner to start fresh with a new script based on the old one
  that only generates the new syntax

- it will avoid regressions to leave the existing code alone while
  we're moving users to the new names

Keep the existing script by moving it to gen_legacy_defines.py and
changing a few comments and strings around. It's responsible for
generating:

- devicetree.conf: only needed by deprecated kconfigfunctions
- devicetree_legacy_unfixed.h: "old" devicetree_unfixed.h macros

Put a new gen_defines.py in its place. It generates:

- zephyr.dts
- devicetree_unfixed.h in the new syntax

Include devicetree_legacy_unfixed.h from devicetree.h so no DT users
are affected by this change.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -969,181 +292,211 @@ def encode_zephyr_multi_level_irq(irq, irq_num):
+    return None
+
+
+def phandle_array_macros(prop, macro2val, macro):
```
I'd return a dictionary instead of passing in a dictionary to extend. (Could use a list of tuples too.)

mbolivar replied:
Will return a dict

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -969,181 +292,211 @@ def encode_zephyr_multi_level_irq(irq, irq_num):
+
+def controller_and_data_macros(entry, prefix, suffix):
+    # Helper procedure shared by phandle_array_macros() and
+    # write_spi_cs_gpio(). The comments before each macro definition
```
Could explain what the helper function does.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -969,181 +292,211 @@ def encode_zephyr_multi_level_irq(irq, irq_num):
-            .replace(".", "_") \
-            .replace("+", "PLUS") \
-            .upper()
+    return re.sub('[-,.@/+]', '_', s.lower())
```
"+" is turned into "PLUS" to give nicer output for stuff like `compatible = "arm,cortex-m0+"`.

mbolivar replied:
> "+" is turned into "PLUS" to give nicer output for stuff like `compatible = "arm,cortex-m0+"`.

Thanks for explaining the usage. I don't think the benefit to the output is worth the extra complication in the documentation, though. We would have to:

- say "special characters are replaced with underscores, *except* for plus, which is replaced with PLUS", and 
- replace "lowercase-and-underscores" with something like "lowercase-and-underscores-except-plus" everywhere in devicetree.h's doxygen strings

ulfalizer replied:
Up to you. I was just copying the old logic.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -277,669 +172,97 @@ def relativize(path):
-    # for enabled nodes. These are flags for which devices exist.
-
+    # Aliases
+    idents.extend(f"N_ALIAS_{str2ident(alias)}" for alias in node.aliases)
```
Could get rid of the `idents = []` assignment and do this:

```python3
idents = [f"N_ALIAS_{str2ident(alias)}" for alias in node.aliases]
```

mbolivar replied:
That's nicer.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -969,181 +292,211 @@ def encode_zephyr_multi_level_irq(irq, irq_num):
 
-def write_spi_dev(node):
-    # Writes SPI device GPIO chip select data if there is any
+        plen = prop_len(prop)
```
Could maybe have an `array_prop()` function that returns true/false instead, and do `if "phandle" not in prop.type and array_prop(prop)`.

Don't have to figure out that `None` length = not array then.

ulfalizer reviewed (commented):
Random comments from skimming.

mbolivar-nordic committed:
drivers: serial: convert uart_nrfx_uart to new DT API

Use the new devicetree.h API instead of the legacy macros.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
drivers: spi: convert spi_nrfx_spim.c to new DT API

This allows us to start using DT_NODELABEL() to access SPIMs that way,
instead of via an alias.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar commented on scripts/dts/gen_defines.py:
```diff
@@ -277,669 +172,97 @@ def relativize(path):
+            continue
+        idents.append(f"N_INST_{instance_no}_{str2ident(compat)}")
+    # Node labels
+    idents.extend(f"N_NODELABEL_{str2ident(label)}" for label in node.labels)
```
> (`idents += [f"N_NODELABEL_{str2ident(label)}" for label in node.labels]` works too.)

Since they both work and there's no reason to change this, I'm leaving it as is.

mbolivar commented on scripts/dts/gen_defines.py:
```diff
@@ -277,669 +172,97 @@ def relativize(path):
 
-def write_code_partition(edt):
-    # Writes output for the node pointed at by zephyr,code-partition in /chosen
+    out_comment("availability and alternate IDs:")
```
> Capital A for consistency maybe (and same for other messages).

I think I got them all

mbolivar commented:
Thanks @ulfalizer; comments addressed.

I've done a couple of test conversions of nRF drivers as well.

mbolivar commented:
> -:2712: WARNING:UNDOCUMENTED_DT_STRING: DT compatible string "vnd,soc-i2c" appears un-documented -- check ./dts/bindings/
> #2712: FILE: doc/guides/dts/main-example.dts:16:
> +			compatible = "vnd,soc-i2c";
> 
> -:2712: WARNING:UNDOCUMENTED_DT_STRING: DT compatible string vendor "vnd" appears un-documented -- check ./dts/bindings/vendor-prefixes.txt
> #2712: FILE: doc/guides/dts/main-example.dts:16:
> +			compatible = "vnd,soc-i2c";

@ulfalizer do you know if we can whitelist "vnd," for examples here?

mnkp commented:
When we use the new macros such as `DT_INST()`, `DT_PROP()` in the device drivers we should be fine. Even if the macros fail to expand correctly the person dealing with the error will be the author of the device driver. I'm however concerned about the usage of `DT_ALIAS()` within the application code. Here, if the macro fails to expand to a known symbol the person dealing with the error message will be a random user who doesn't have to be very familiar with Zephyr.

I propose to convert the `samples/basic/button` sample app (it's a bit more complex than `samples/basic/blinky` but still fairly straightforward) to use `DT_ALIAS()` and then try to compile it on a board that doesn't have `aliases` node. It's something a random, new Zephyr user is likely to do.

Then we should make a judgment, trying to asses if the error message produced by the compiler is something this user is able to deal with.

I run some tests locally, but it looks like we are missing macros to access properties with type phandle-array, i.e. extract pin value from gpios property.
```
led0: led_0 {
	gpios = <&gpio0 13 GPIO_ACTIVE_LOW>;
	label = "Green LED 0";
};
```
so I'm not really sure what the end result would be.

Zephyr's logging subsystem is also using macros extensively and the error messages produced are bulky and complex, difficult to interpret.

galak commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -362,11 +362,11 @@ static int spim_nrfx_pm_control(struct device *dev, u32_t ctrl_command,
 
-#define SPIM_NRFX_MISO_PULL_DOWN(idx) \
-	IS_ENABLED(DT_NORDIC_NRF_SPIM_SPI_##idx##_MISO_PULL_DOWN)
+#define SPIM(idx) DT_NODELABEL(spi##idx)
```
Do you know why this is NODELABEL?

mbolivar replied:
> drivers may use SoC specific NODELABEL identifiers to refer to devices when they become available (#22555), e.g. if HAL constraints prevent use of INST 

That's from a proposal we discussed while you were away and hope you find agreeable.

https://github.com/zephyrproject-rtos/zephyr/issues/19285#issuecomment-591099718

galak replied:
I guess my question is why is this NODELABEL vs INST?

mbolivar replied:
https://github.com/zephyrproject-rtos/zephyr/issues/19285#issuecomment-591092481 is linked to from the proposal

mbolivar replied:
just copy/pasting:

at least one place where this is an invalid assumption:

    Here the solution [to use DT_INST_X] is fine because user code will access specific devices through the instance identifier (label property value) and the driver doesn't care which one is which.

In particular, the Nordic HAL exposes some ISR routines for SPI peripherals through names that correspond to their SoC IDs, namely nrfx_spim_X_irq_handler:

https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/spi/spi_nrfx_spim.c#L396

I.e. the driver does care which one is which in some cases, even when drivers have identical compats, so drivers should be able to use NODELABEL in time of need.

It also seems to me there's no disagreement to #22555 (generate NODELABEL node identifiers) in principle, there's just desire to settle on a more robust naming convention ( #22964) for DT macros before adding more names.

I will follow up with another comment summarizing the discussion and making a proposal for how we can close this.

mbolivar-nordic committed:
samples: convert blinky to new DT API

This doesn't sacrifice any readability when compiled for boards that
don't support this alias.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar commented:
> I propose to convert the `samples/basic/button` sample app (it's a bit more complex than `samples/basic/blinky` but still fairly straightforward) to use `DT_ALIAS()` and then try to compile it on a board that doesn't have `aliases` node. It's something a random, new Zephyr user is likely to do.

Sounds like a great ideal.

galak committed:
drivers: serial: uart_sifive: convert to new DT API

Use the new devicetree.h API instead of the legacy macros.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar commented:
> > I propose to convert the `samples/basic/button` sample app (it's a bit more complex than `samples/basic/blinky` but still fairly straightforward) to use `DT_ALIAS()` and then try to compile it on a board that doesn't have `aliases` node. It's something a random, new Zephyr user is likely to do.
> 
> Sounds like a great idea.

Now done.

mbolivar-nordic committed:
doc: dts: revisit documentation

This is joint work with Kumar Gala (see signed-off-by).

Document the changes to the generated node macros in macros.bnf,
moving the old file to legacy-macros.bnf and putting it in its own
section.

The actual generated macros are now a low-level detail, so rewrite the
foregoing sections as examples in terms of the new <devicetree.h> APIs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

erwango commented on drivers/serial/uart_nrfx_uart.c:
```diff
@@ -932,18 +930,17 @@ static int uart_nrfx_init(struct device *dev)
-	IRQ_CONNECT(DT_NORDIC_NRF_UART_UART_0_IRQ_0,
-		    DT_NORDIC_NRF_UART_UART_0_IRQ_0_PRIORITY,
+	IRQ_CONNECT(UART_0_IRQ,
+		    DT_IRQ(UART_0, priority),
```
Mixing direct calls and macros (such as UART_0_IRQ) in small portion of code feels confusing.
Might be better to avoid that

mbolivar replied:
> Mixing direct calls and macros (such as UART_0_IRQ) in small portion of code feels confusing.

Sure. I was avoiding defining a macro and only using it once, but I agree with you.

jfischer-no reviewed (commented):
+1

jfischer-phytec-iot commented:
>* Do we make the macro's Z_DT_INST_ or just DT_INST_
  (for example Z_DT_INST_PROP() or DT_INST_PROP())

What would be the reason for `Z_DT_ `instead just `DT_`? I am happy that this macros helps to fix a lot of style issues and think they should be as short as possible, also just `DT_`

galak commented:
> What would be the reason for `Z_DT_ `instead just `DT_`? I am happy that this macros help to fix a lot of style issues and think they should be as short as possible, also just `DT_`

No great reason, I thought it would be useful for grepping differences for `DT_INST` v `Z_DT_INST` but I'm good with going with just `DT_`

galak committed:
drivers: sensor: fxos8700: convert to new DT API

Use the new devicetree.h API instead of the legacy macros.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar-nordic committed:
scripts: dts: edtlib: add EDT.chosen_nodes property

This returns the entire logical {name: Node} dictionary which is
currently being accessed element by element via chosen_node(name).

It will be used in a new gen_defines.py for moving the handling of
chosen nodes into C from Python.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

galak committed:
edtlib: allow register addr / size to be None

If the #address-cells property for a register is 0 than we set the addr
value of the reg to None.  Similar, if #size-cells is 0 than we set the
size value to None for the reg.

Fixup kconfigfunctions.py to handle reg.size and reg.addr being None.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
gen_defines.py: rename gen_defines.py to gen_legacy_defines.py

This rename is mostly to easy git managment and review so any changes or
the addition of the new gen_defines.py doesn't look like a diff against
the old code if you look at just that commit.

We keep changes to a minimum to just keep things building with the
renamed gen_legacy_defines.py.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -179,22 +185,33 @@ def get_node(self, path):
+                # DTS value is not phandle or string, or path doesn't exist
+                continue
+
+            if node in self._node2enode:
```
`_node2enode` includes all nodes, so no need to check.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -179,22 +185,33 @@ def get_node(self, path):
 
-        # to_path() checks that the node exists
-        return self._node2enode[chosen.props[name].to_path()]
+    def chosen_node(self, name):
```
I'd skip this function, since it's trivial to implement.

mbolivar-nordic replied:
What do you mean? This function was here before and it's still needed by the legacy script.

ulfalizer replied:
Could patch the callers to do

```python
edt.chosen_nodes.get("foo")
```

instead of

```python
edt.chosen_node("foo")
```

Nice to not have functions that are overly trivial I think. Makes it easier to find all the spots where `chosen_nodes` is used too.

mbolivar-nordic replied:
Yes, we could, but I am trying to limit changes to gen_legacy_defines.py to avoid regressions. Basically trying to freeze that file as mentioned in the top level comment:

```
# This script is similar to gen_defines.py, but is for the legacy
# macro syntax used in e.g. Zephyr 2.2.
#
# It should be considered frozen code.
```

Making this change will in turn cause other "cleanups" to this file to be the next logical course of action (https://gist.github.com/mbolivar-nordic/6dc036905ba9ebb0cb956b5861b00725), because once you have a way to get all the chosen nodes, it's wasted cycles to recompute them repeatedly.

So I'm wondering why we can't just remove this method once we remove gen_legacy_defines.py?

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -1245,8 +1262,14 @@ def _init_regs(self):
-            reg.addr = _translate(to_num(raw_reg[:4*address_cells]), node)
-            reg.size = to_num(raw_reg[4*address_cells:])
+            if address_cells == 0:
+                reg.addr = None
```
The documentation for the `Register` class needs to be updated.

mbolivar-nordic replied:
done

ulfalizer commented:
@galak 

Would be nice to have a motivation for this change in the commit message. "This allows yada yada..." or the like.

```
edtlib: allow register addr / size to be None

If the #address-cells property for a register is 0 than we set the addr
value of the reg to None.  Similar, if #size-cells is 0 than we set the
size value to None for the reg.

Fixup kconfigfunctions.py to handle reg.size and reg.addr being None.
```

galak commented on include/devicetree.h:
```diff
@@ -5,10 +5,1273 @@
+ * @return the label property for the referenced node at index idx
+ * @see DT_PHA_PHANDLE_IDX()
+ */
+#define DT_GPIO_LABEL_IDX(node_id, gpio_pha, idx) \
```
We should pull the DT_GPIO macro's out into either into include/drivers/gpio.h or create a new devicetree header group and put this in include/devicetree/gpio.h.  The reason is we'll need similar macro's for IO_CONTROLLERs, CLOCKS, and any other phandle/phandle-array.  @mnkp / @pabigot any thoughts?

carlescufi replied:
API meeting:

subsystem-specific DT helper macros belong either in:

`include/devicetree/<foo>.h`
or
`include/drivers/<foo>.h`

If `include/devicetree.h` includes `include/devicetree/*` then the application only ever needs to include `<devicetree.h>`.

Quorum: `include/devicetree/*`

mbolivar-nordic commented:
> -:2715: WARNING:UNDOCUMENTED_DT_STRING: DT compatible string "vnd,soc-i2c" appears un-documented -- check ./dts/bindings/
> #2715: FILE: doc/guides/dts/main-example.dts:22:
> +			compatible = "vnd,soc-i2c";
> 
> -:2715: WARNING:UNDOCUMENTED_DT_STRING: DT compatible string vendor "vnd" appears un-documented -- check ./dts/bindings/vendor-prefixes.txt
> #2715: FILE: doc/guides/dts/main-example.dts:22:
> +			compatible = "vnd,soc-i2c";

@ulfalizer is there any way to whitelist this? should I patch ci-tools?

anangl commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -366,11 +366,17 @@ static int spim_nrfx_pm_control(struct device *dev, u32_t ctrl_command,
-#define SPIM_NRFX_MISO_PULL_DOWN(idx) \
-	IS_ENABLED(DT_NORDIC_NRF_SPIM_SPI_##idx##_MISO_PULL_DOWN)
+/*
+ * We use NODELABEL here because the nrfx HAL API requires us to call
```
HAL can be misleading in the context of nrfx, because there we describe this way a layer for accessing peripheral registers via inline functions. And the functions for which we need the SoC peripheral instance index are in nrfx drivers API. I'd suggest to use just "nrfx API" instead.

erwango commented on drivers/serial/uart_sifive.c:
```diff
@@ -8,6 +8,8 @@
  * @brief UART driver for the SiFive Freedom Processor
  */
 
+#define DT_DRV_COMPAT sifive_uart0
```
Any way we can we use the real compatible string ("sifive,uart0") here ?
If not, I'd actually prefer the caps version. Adding a new, intermediate (with "_" but lower case) modified string is confusing IMHO.

pabigot replied:
I believe this must be lexically a C identifier fragment so it can be part of a macro identifier, so using a comma is out.

AFAICT the only SiFive-related compatible that is not consistently lower-case is the SOC `SiFive,FE310G-0002-Z0-soc`. which IMO is incorrectly named not least because it's not the same as the other SiFive compatibles.

I'd support preserving case in the compatible string as long as the devicetree standard and/or reference implementation dtc considers case significant when checking compatibility.  If non-Zephyr devicetree considers case insignificant then I'd support canonicalizing to the lower-case.

erwango replied:
As long as we don't use the exact string, I don't see the point to keep the exact same case.
So, from the moment we have to convert the string, I'd say `SIFIVE_UART0` would be easier to grep (and potentially find related generated macros)

mbolivar-nordic replied:
> Any way we can we use the real compatible string ("sifive,uart0") here ?

I definitely don't have the preprocessor skills to make that happen. Not sure it's even possible, really.

The generated header looks like this:

```
#define DT_N_INST_0_sifive_uart0 DT_N_S_soc_S_serial_10013000

/* Generic property macros: */
#define DT_N_S_soc_S_serial_10013000_P_reg {268513280 /* 0x10013000 */, 4096 /* 0x1000 */}
#define DT_N_S_soc_S_serial_10013000_P_reg_EXISTS 1
#define DT_N_S_soc_S_serial_10013000_P_interrupts {3 /* 0x3 */}
#define DT_N_S_soc_S_serial_10013000_P_interrupts_EXISTS 1
#define DT_N_S_soc_S_serial_10013000_P_clock_frequency 16000000
#define DT_N_S_soc_S_serial_10013000_P_clock_frequency_EXISTS 1
... etc
```

So the macro expansion is roughly `DT_DRV_INST(0)` -> `DT_INST(0, DT_DRV_COMPAT)` -> `DT_INST(0, sifive_uart0)` -> `DT_N_INST_0_sifive_uart0` -> `DT_N_S_soc_S_serial_10013000`.

The final token `DT_N_S_soc_S_serial_10013000` can get pasted onto `_P_clock_frequency` or something like that to get property values, etc.

So we're relying on being able to paste DT_DRV_COMPAT in the preprocessor, like `DT_N_INST_##inst##_##DT_DRV_COMPAT`, for this all to work. That means it has to be a C token.

So using "sifive,uart0" would require a solution where we could convert that string into that token at preprocessor time. I don't know how to do that.... any ideas?

> If not, I'd actually prefer the caps version. Adding a new, intermediate (with "_" but lower case) modified string is confusing IMHO.

The underlying macros require that names from devicetree are lowercased in order to avoid possible namespace clashes. Reversing that so they are uppercase is possible, but would be a lot of work at this point. Is this a strong preference?

mbolivar-nordic replied:
This is the same requirement introduced for labels, properties, etc, btw -- special characters are consistently converted to underscores everywhere.

mbolivar-nordic replied:
> Reversing that so they are uppercase is possible, but would be a lot of work at this point. Is this a strong preference?

To clarify, we would do this everywhere: properties would also need to be uppercased, etc. And the generated macro names would have to use lowercase, so it would look like `DT_n_s_SOC_s_SERIAL_10013000_p_CLOCK_FREQUENCY` in the generated file.

IMO this is a worse option because devicetree identifiers tend to be lowercase most of the time, so people just need to know to convert special characters to underscores in the C API.

We do need a consistent case in this naming scheme to avoid collisions. For example, uppercase _P_ is used to separate the node identifier from the property name. Since the identifier and property names are lowercase, there is no ambiguity. If we don't do a case transformation, we can't be sure.

Please see macros.bnf for the details.

erwango replied:
@mbolivar-nordic, thanks for the detailed explanation. Ack to say it's minor a drawback compared to what the global solution conveys

erwango replied:
@mbolivar-nordic, thanks for the detailed explanation. Ack to say it's minor a drawback compared to what the global solution conveys

erwango commented on drivers/serial/uart_sifive.c:
```diff
@@ -8,6 +8,8 @@
  * @brief UART driver for the SiFive Freedom Processor
  */
 
+#define DT_DRV_COMPAT sifive_uart0
```
Also, what if the driver supports two compatibles? 
Are there some usage rules?

mbolivar-nordic replied:
> Also, what if the driver supports two compatibles?
> Are there some usage rules?

There are some limitations in this regard. https://github.com/zephyrproject-rtos/zephyr/issues/19904

mbolivar-nordic replied:
That being said, I think at an API level there's nothing stopping a driver from making their own instance wrappers for each compatible, once the underlying issues with the bindings are resolved. The `DT_N_INST_i_compat` macro namespaces will not need to change.

erwango replied:
I have several cases in mind, need deeper study to see impacts

erwango commented on drivers/serial/uart_sifive.c:
```diff
@@ -8,6 +8,8 @@
  * @brief UART driver for the SiFive Freedom Processor
  */
 
+#define DT_DRV_COMPAT sifive_uart0
```
This being said, I think I like it.

mbolivar-nordic commented:
@anangl comments addressed, thank you for your thorough review.

mbolivar-nordic committed:
dts: bindings: add fake 'vnd' vendor

This is useful for devicetree documentation, examples, and tests,
where we need to put something for the vendor but we can't use an
actual piece of hardware for some reason.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
This has failed `tests/kernel/context` for qemu_cortex_r5 twice. I think that's the usual false negative related to timers in shippable not working. I ran it locally and it works fine.

pabigot commented on include/devicetree.h:
```diff
@@ -1,15 +1,1283 @@
+ * _NAME_<name>_VAL_<val>: phandle array's property specifier by name
+ * _NAME_<name>_VAL_<val>_EXISTS: cell value exists, by name
+ * _NAME_<name>_PH: phandle array's phandle by name
+ * _ENUM_IDX: property's value as an index into bindings enum
```
It would be easier to search for conflicts (and interpretation) if these were in alphabetical order.

pabigot commented on include/devicetree.h:
```diff
@@ -1,15 +1,1283 @@
+ *   {"hello", "world"}, etc.)
+ * - phandle: a node identifier
+ *
+ * For other properties, behavior is undefined.
```
Ack; per #21273 some future support for enum values would be nice.

pabigot commented on include/devicetree.h:
```diff
@@ -1,15 +1,1283 @@
+ * These properties are handled as special cases:
+ *
+ * - reg property: use DT_REG_HAS_IDX(node_id, idx) instead
+ * - interrupts property: use DT_IRQ_HAS_IDX(node_id, idx) instead
```
Is it "an error to use this macro with the above properties." as with `DT_PROP_LEN`?

mbolivar-nordic replied:
Yes. Will add.

pabigot commented on include/devicetree.h:
```diff
@@ -1,15 +1,1283 @@
+ * uint8-array, or phandles, this expands to the idx-th array element
+ * as an integer, string literal, or node identifier respectively.
+ *
+ * It's not safe to use this macro with the special "reg" or
```
This is the third way a special case for `reg` and `interrupts` is identified and described.  That's not wrong, but some people might see the differences and wonder if they're significant....

mbolivar-nordic replied:
I think it's wrong. Will fix; thanks.

pabigot commented on include/devicetree.h:
```diff
@@ -1,15 +1,1283 @@
+ *
+ * - exists,
+ * - is enabled (has status property "okay"), and
+ * - has a matching binding
```
Is the test conjunctive or disjunctive of these conditions?  Presumably conjunctive, since it can't be enabled if it doesn't exist, and IIRC is an error to have a node that doesn't correspond to a binding (or is that changed now, so we can add arbitrary stuff without validation?).  In any case this seems a little confusing, especially since the implementation seems to only reference "exists".

Another point for the future: per #19448 it's likely that very soon we're going to need to be able to get information about nodes where status is not "okay" but one of the values that indicates it can be enabled at runtime.

mbolivar-nordic replied:
> Is the test conjunctive or disjunctive of these conditions? Presumably conjunctive

Conjunctive is correct: "exists, is enabled [...], and". Will try to clarify. Pay no attention to the macro name behind the curtain ;).

> Another point for the future: per #19448 it's likely that very soon we're going to need to be able to get information about nodes where status is not "okay" but one of the values that indicates it can be enabled at runtime.

Seems like this can be supported at that time via a new `DT_NODE_HAS_STATUS(node_id, SOME_SPECIAL_TOKEN)`.

pabigot commented on doc/guides/dts/howtos.rst:
```diff
@@ -1,479 +1,546 @@
-Other examples of board-specific customization is pointing properties in
-``aliases`` and ``chosen`` to the right nodes (see :ref:`dt-alias-chosen`), and
-making GPIO/pinmux assignments.
+.. code-block::
```
This causes a missing argument diagnostic during build, needs:

    .. code-block:: c

mbolivar-nordic replied:
Interesting; I didn't get one. Will add anyway.

pabigot reviewed (commented):
Some initial comments, nothing critical.  I still need to read through the user documentation.  On the whole I think this'll be a great improvement.

galak commented on include/devicetree.h:
```diff
@@ -362,9 +362,9 @@
- * phandle (which is just a u32_t node property, after all) and a "real"
- * struct device.
+ * These are special-cased to manage the impedance mismatch between
+ * phandles, whiare just a u32_t node properties that only make sense
```
typo here `whiare`

galak commented on doc/guides/dts/api-usage.rst:
```diff
@@ -0,0 +1,384 @@
+One common use case for phandle properties is referring to other hardware in
+the tree. In this case, you usually want to convert the devicetree-level
+phandle to a Zephyr driver-level :ref:`struct device <device_model_api>`. You
+can use :c:func:`DT_PH_LABEL_IDX` or :c:func:`DT_PH_LABEL` for this depending
```
We don't have DT_PH_LABEL_IDX or DT_PH_LABEL.  what should these refs be?

mbolivar-nordic commented:
@pabigot @galak comments addressed

galak commented:
I think this will fix #21369 - @pabigot ?

galak commented:
Also I think this address #19285 (@pabigot )

mbolivar-nordic committed:
doc: release-notes-2.3: add devicetree changes

Mention the new devicetree.h API and link to relevant
resources.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

pabigot commented:
> Also I think this address #19285 (@pabigot )

Yes, this is enough to cover #19285 and #21369.  I was going to update the description to add them, but it doesn't seem to be using the GFM-style close directives.

mbolivar-nordic commented:
>  I was going to update the description to add them, but it doesn't seem to be using the GFM-style close directives.

???

```
Fixes: #22964
Fixes: #22555
Fixes: #22554
```

There they are...

mbolivar-nordic committed:
CODEOWNERS: devicetree updates

Update maintainers for code and docs.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

pabigot commented:
When I tried to edit the description I saw (see):
```
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/22964
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/22555
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/22554
```
which I'm not sure github is smart enough to handle (does it interpret the URL to be the same as `#22965`).  [This](https://help.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords) only promises it for the issue number.  We'll find out....

fabled commented:
Perhaps the new API could also address how to create templated device structure instances so that header file generation is not needed. See e.g. `drivers/serial/uart_ns16550_port_x.h`.

mbolivar-nordic commented:
> Perhaps the new API could also address how to create templated device structure instances

I may be misunderstanding you, but I think not, just because devicetree.h != device.h, and device.h is where struct device is defined.

I believe you're maybe talking about https://github.com/zephyrproject-rtos/zephyr/issues/22941? That is critical work and a logical next step, but, in my opinion, something to build on top of devicetree.h rather than inside of it from a layering perspective.

> so that header file generation is not needed. See e.g. `drivers/serial/uart_ns16550_port_x.h`.

That file looks like a cmake `configure_file()` input. I don't follow why "header file generation is not needed". That just looks like yet another templating strategy on top of what people have variously suggested: [chaos](https://github.com/ldionne/chaos-pp/blob/master/chaos/preprocessor.h), [jinja](https://jinja.palletsprojects.com/en/2.11.x/), vanilla C preprocessor, etc. The information still has to come from a particular devicetree somehow, so there's some generation.

pabigot reviewed (approved):
IMO this is far enough along to commit to it and get it into master so we can start using it, so +1 from me.

I would like restoration of the ability to reference initializer lists for phandle array elements ASAP.  So I can do:

```
-       .bus_name = DT_INST_0_MAXIM_DS3231_BUS_NAME,
+       .bus_name = DT_INST_BUS_LABEL(0),
        /* Driver does not currently use 32k GPIO. */
-#ifdef DT_INST_0_MAXIM_DS3231_ISW_GPIOS_CONTROLLER
-       .isw_gpios = DT_INST_0_MAXIM_DS3231_ISW_GPIOS,
+#if DT_INST_NODE_HAS_PROP(0, isw_gpios)
+        .isw_gpios = DT_INST_GPIO(0, isw_gpios),
 #endif
-       .addr = DT_INST_0_MAXIM_DS3231_BASE_ADDRESS,
+       .addr = DT_INST_REG_ADDR(0),
```
Right now I have to hack to get that feature back:

```
#define DT_INST_GPIO(inst, gpio_pha) {		\
	DT_INST_GPIO_LABEL(inst, gpio_pha),	\
	DT_INST_GPIO_PIN(inst, gpio_pha),	\
	DT_INST_GPIO_FLAGS(inst, gpio_pha),	\
}
```

It would probably be appropriate to do that simultaneously with defining standard structures to hold the phandle data, e.g.:

```
struct dt_node_gpio {
    const char *controller;
    gpio_pin_t pin;
    gpio_dt_flags_t flags;
};
```

anangl reviewed (approved):
LGTM. nRF related changes are fine. A few imperfections in documentation spotted.

mbolivar commented on doc/guides/dts/legacy-macros.rst:
```diff
@@ -819,14 +824,327 @@ for each partition, derived from ``reg``. The ``*_OFFSET`` and ``*_SIZE``
+   #define FLASH_AREA_DEF_NAME_OFFSET_0     0xSTART_OFFSET
+   #define FLASH_AREA_DEF_NAME_SIZE_0       0xSIZE
+   #define FLASH_AREA_DEF_NAME_OFFSET       FLASH_AREA_MCUBOOT_OFFSET_0
+   #define FLASH_AREA_DEF_NAME_SIZE         FLASH_AREA_MCUBOOT_SIZE_0
```
Interesting, I guess this was a issue before this PR as I'm just moving these lines around. Will fix, though.

mbolivar commented on doc/guides/dts/howtos.rst:
```diff
@@ -1,479 +1,548 @@
+    */
 
-.. code-block:: none
+   struct some_api my_api_funcs = { /* ... */ };
```
I'm going to change the below to my_api_funcs instead, thanks.

mbolivar commented:
@anangl comments addressed, thanks for the second round of review.

mbolivar commented:
> I would like restoration of the ability to reference initializer lists for phandle array elements ASAP. So I can do:

This was originally present but was removed when we moved towards generating node identifiers in place of phandles.

For GPIOs, we could do something like add the following convenience:

`DT_GPIO_LPF_BY_IDX(node, gpios, i) --> {"a-label-property", some_pin_number, some_flags_number}.

Where LPF == label, pin, and flags.

pabigot commented:
> > I would like restoration of the ability to reference initializer lists for phandle array elements ASAP. So I can do:
> 
> This was originally present but was removed when we moved towards generating node identifiers in place of phandles.

OK, that's something to discuss then, because a phandle+specifier is a very useful thing to be able to treat as a single unit, for GPIO, IO channels, probably everything.  `LPF` as a name to capture the GPIO phandle+specifier seems opaque.  For now I'll do my own within each module where I need it.

galak commented:
> > > I would like restoration of the ability to reference initializer lists for phandle array elements ASAP. So I can do:
> > 
> > 
> > This was originally present but was removed when we moved towards generating node identifiers in place of phandles.
> 
> OK, that's something to discuss then, because a phandle+specifier is a very useful thing to be able to treat as a single unit, for GPIO, IO channels, probably everything. `LPF` as a name to capture the GPIO phandle+specifier seems opaque. For now I'll do my own within each module where I need it.

@pabigot - Do we have an issue open for this?

pabigot commented:
> @pabigot - Do we have an issue open for this?

No, because there's a working solution for the old API.  The new API doesn't have a replacement.

mbolivar commented:
> OK, that's something to discuss then, because a phandle+specifier is a very useful thing to be able to treat as a single unit, for GPIO, IO channels, probably everything.

I agree, but I think the correct solution is for the initializer generated for a phandle+specifier to replace the phandle with a pointer to the struct device, which we can't do yet.

pabigot commented:
> I agree, but I think the correct solution is for the initializer generated for a phandle+specifier to replace the phandle with a pointer to the struct device, which we can't do yet.

In the final configuration, yes.  For now, I need something to replace the existing uses of initializer-lists that include the name of the device along with the specifier.  I will deal with that on a case-by-case basis.

galak commented:
> > I agree, but I think the correct solution is for the initializer generated for a phandle+specifier to replace the phandle with a pointer to the struct device, which we can't do yet.
> 
> In the final configuration, yes. For now, I need something to replace the existing uses of initializer-lists that include the name of the device along with the specifier. I will deal with that on a case-by-case basis.

How do you feel about a macro per subsystem/class?  Or do you see the need for this for phandle/phandle-array out of tree?

mbolivar-nordic commented:
> How do you feel about a macro per subsystem/class? Or do you see the need for this for phandle/phandle-array out of tree?

Can we just do this as needed for now since it's temporary? Any other cases besides GPIO?

I really want to be able to turn a phandle into a struct device* sooner rather than later, so I am hoping we don't have to spend too much review / discussion bandwidth solving an interim problem...

pabigot commented:
There are other cases, yes, ADC in particular.  But external workarounds are fine for now.  #23691 can be resolved in the context of generating a device handle instead of the name; that's preferable anyway.
